src/ripple/overlay/MessageScheduler.h

```diff
@@ -0,0 +1,496 @@
+    // IDs, `PeerImp` can dispatch incoming responses `TMLedgerData` messages to
+    // the correct receiver. Similarly for the `seq` field on
+    // `TMGetObjectByHash` messages.
+    static constexpr RequestId MINIMUM_REQUEST_ID = 1 << 24;
+
+private:
+    using Clock = std::chrono::steady_clock;
+    using Timer = boost::asio::basic_waitable_timer<Clock>;
+    using Time = std::chrono::time_point<Clock>;
+
+    // An in-flight request.
+    struct Request
+    {
+        RequestId id;
+        Channel channel;
+        Receiver* receiver;
+        Timer timer;
+        Time sent;
+    };
+
+    boost::asio::io_service& io_service_;
+    beast::Journal journal_;
+
+    // Offers are negotiations between peers and senders.
+    // This mutex must be locked when handling either set.
+    std::mutex offersMutex_;
+    std::vector<Channel> channels_;
+    // TODO: Use a priority queue.
+    std::vector<Sender*> senders_;
+    bool stopped_ = false;
+
+    // Randomize the first ID to avoid collisions after a restart.
+    // C++23: We may be able to switch to std::randint.
+    // https://en.cppreference.com/w/cpp/experimental/randint
+    std::atomic<RequestId> nextId_ =
+        rand_int<RequestId>(MINIMUM_REQUEST_ID, 1 << 24);
+    std::mutex requestsMutex_;
+    // TODO: Might make sense to use a set instead.
+    hash_map<RequestId, std::unique_ptr<Request>> requests_;
+
+public:
+    MessageScheduler(
+        boost::asio::io_service& io_service,
+        beast::Journal const& journal)
+        : io_service_(io_service), journal_(journal)
+    {
+    }
+
+    /**
+     * If there are any waiting senders, offer these channels to them.
+     * Add any remaining open channels to the pool.
+     *
+     * TODO: Maybe this can be private.
+     */
+    void
+    connect(std::vector<Channel> channels);
+
+    void
+    connect(std::shared_ptr<Peer> peer)
+    {
+        JLOG(journal_.trace())
+            << "connect,id=" << peer->id()
+            << ",address=" << peer->getRemoteAddress().to_string();
+        Channel channel{peer->id(), peer};
+        // TODO: Let a peer choose its number of channels when it connects.
+        constexpr std::size_t CHANNELS_PER_PEER = 1;
+        std::vector<Channel> channels(CHANNELS_PER_PEER, channel);
+        return connect(channels);
+    }
+
+    /**
+     * If this peer has any channels in the pool, remove them.
+     * If it is responsible for any in-flight requests,
+     * call their failure callbacks.
+     * If those callbacks schedule any new senders,
+     * offer them the other channels in the pool.
+     */
+    void
+    disconnect(Peer::id_t peerId);
+
+    /**
+     * @return True if the sender was scheduled, guaranteeing that one of its
+     * callbacks will be called. False if the sender was immediately
+     * discarded (because the scheduler has stopped), returning responsibility
+     * for its lifetime to the caller.
+     */
+    bool
+    schedule(Sender* sender);
```

Comment:
This doesn't seem particularly difficult to me. If a `Sender` needs to manage its own lifetime (not all do), then it should delete itself at the end of `onDiscard`, or at the end of `onOffer` if it closed a channel and did not reschedule itself. If a `Receiver` needs to manage its own lifetime (not all do), then it should delete itself at the end of `onSuccess` and `onFailure`.

I'm not a fan of reference-counting everything. Why not just introduce a garbage collector if we never want to spend two seconds thinking about lifetimes?

Why would `MessageScheduler` store a `Receiver`? That's never coming. Neither will it store a `Sender`.

Reply:
What I've observed, with myself and others, is that lifetime management is usually not hard when you think about it.  But folks often don't think about it.  For example, I've seen myself and others design plenty of things without consideration of what happens at shutdown.  Then the app randomly crashes at shutdown.  I found and fixed a number of these crash-on-shutdown bugs in rippled as part of my tenure working at Ripple.  When possible I fixed them so further code changes would not introduce similar bugs.

Another example: eleven years after range-based for loops were introduced into the language, C++23 changed the language to fix a lifetime problem that newbies and experienced developers both stumbled into: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2644r1.pdf

The problem with lifetime issues is that they are seldom a _primary_ concern when a bug is being fixed or a feature is being added.  You're thinking about the problem at hand.  You're not thinking about lifetime consequences.

> I'm not a fan of reference-counting everything. Why not just introduce a garbage collector if we never want to spend two seconds thinking about lifetimes?

The reason not to do it with a garbage collector is that garbage collectors are (typically) not deterministic.  That's why we use RAII.  A lot.

There's a reason that `std::shared_ptr`, `std::unique_ptr`, and RAII are so ubiquitous in C++.  They solve lifetime issues so the developer can dedicate their brain space to solving problems, not worrying about (or forgetting to worry about) the lifetime consequences of decisions that they have just made.

I'm not saying "hell no".  But I think that if we can design internal software so it is easy to use correctly and hard to misuse then we will have fewer bugs.  I think any step away from that position should require strong _measured_ performance gains.

Just my thoughts on the matter.  I'm open to further discussion.  And mine is not the only voice to consider.

Reply:
I _have_ thought about lifetimes, carefully, for quite a bit of time. That's how I arrived at this design, and why I dedicated an entire section in the documentation to it.

Reply:
I think we're talking past each other.  I really do believe that _you_ have thought hard about lifetimes.  And I further believe, since _you_ have internalized the lifetime requirements, that whenever _you_ write code that uses  `MessageScheduler` you will obey those lifetime requirements.

My concern is about usability by _others_.  Other people _may_ (or may not) read the documentation.  Even after they read it they _may_ forget about the lifetime requirements while coding and debugging.  Others will walk into a pre-existing piece of code to make a modification and not even know that there is documentation that they should read.  All these folks have a fair chance of introducing bugs.

My take is that you have designed this software with speed in mind.  That's laudable.  However, my take at the moment is that it's probably prudent to sacrifice some of that speed to make the `MessageScheduler` easy for _others_ to use correctly and hard for _others_ to misuse.

Writing all of the rules for usage in the documentation does not make it hard for _others_ to misuse, even though I do appreciate the effort.

But we may just end up disagreeing on this point.  Or maybe you can convince me that it actually _is_ hard to misuse `MessageScheduler`.  I'll stop now and let others weigh in.

Reply:
Can anyone sketch an alternative design that implements the changes they want to see? Can we collaborate from that position? Right now I'm just hearing "no, don't do that way" and not "do it this way instead".

I anticipate most users will not be writing their own non-singleton `Sender`s. They'll call or inherit from implementations that execute specific messaging patterns: deluge all available peers with as many requests as possible; or send at most M simultaneous messages until reaching all N peers, stopping at the first response; or broadcast 1 message to every peer as quickly as possible, expecting no response.

This is like a low-level library. The language offers matched constructors and destructors, and it is up to the programmer to implement RAII correctly. Can we never expect the same from our programmers?